dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -25,8 +36,95 @@ namespace ErrorCodes
 {
 extern const int BAD_ARGUMENTS;
 extern const int NOT_IMPLEMENTED;
+extern const int CANNOT_INSERT_NULL_IN_ORDINARY_COLUMN;
 } // namespace ErrorCodes
 
+namespace
+{
+template <typename T>
+consteval bool checkIfSimpleNumericType()
+{
+    return std::is_integral_v<T> || std::is_floating_point_v<T>;
+}
+
+template <typename T>
+consteval bool checkIfDecimalFieldType()
+{
+    return std::is_same_v<T, DecimalField<Decimal32>> || std::is_same_v<T, DecimalField<Decimal64>> || std::is_same_v<T, DecimalField<Decimal128>> || std::is_same_v<T, DecimalField<Decimal256>>;
+}
+
+// When T is Decimal, we should convert it to DecimalField type
+// as we need scale value when executing the comparison operation.
+template <typename T>
+struct ActualCmpDataType
+{
+    using Type = std::conditional_t<checkIfSimpleNumericType<T>(), T, DecimalField<T>>;
+};
+
+template <typename T>
+typename ActualCmpDataType<T>::Type getValue(const ColumnPtr & col_ptr, size_t idx)
+{
+    return (*col_ptr)[idx].get<typename ActualCmpDataType<T>::Type>();
+}
+
+template <typename T, bool is_start, bool is_desc>
+bool isInRangeCommonImpl(T current_row_aux_value, T cursor_value)
+{
+    if constexpr ((is_start && is_desc) || (!is_start && !is_desc))
+        return cursor_value <= current_row_aux_value;
+    else
+        return cursor_value >= current_row_aux_value;
+}
+
+template <bool is_start, bool is_desc>
+bool isInRangeIntImpl(Int64 current_row_aux_value, Int64 cursor_value)
+{
+    return isInRangeCommonImpl<Int64, is_start, is_desc>(current_row_aux_value, cursor_value);
+}
+
+template <typename AuxColType, typename OrderByColType, bool is_start, bool is_desc>
+bool isInRangeNotIntImpl(AuxColType current_row_aux_value, OrderByColType cursor_value)
+{
+    Float64 current_row_aux_value_float64;
+    Float64 cursor_value_float64;
+
+    if constexpr (checkIfDecimalFieldType<AuxColType>())
+        current_row_aux_value_float64 = current_row_aux_value.getValue().template toFloat<Float64>(current_row_aux_value.getScale());
```

Comment:
Why decimal value need to be converted to float?

Reply:
> Why decimal value need to be converted to float?

When these two data type are all decimal, it's no need to convert then to float, I will fix it. However, when one of the data type is float, converting decimal to float is necessary.

Reply:
I am a little confused, decimal should be the parent-type of float, will converting decimal to float not lose precision?